runtime/src/topo/hwloc/topo-hwloc.c

```diff
@@ -298,15 +309,40 @@ void getNumCPUs(void) {
   numCPUsLogAcc = hwloc_bitmap_weight(logAccSet);
   CHK_ERR(numCPUsLogAcc > 0);
 
-  hwloc_bitmap_free(logAccSet);
-
   //
   // all PUs
   //
   numCPUsLogAll = hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_PU);
   CHK_ERR(numCPUsLogAll > 0);
 }
 
+//
+// Fills the "cpus" array with the hwloc "cpuset" (a bitmap whose bits are
+// set according to CPU physical OS indexes).
+//
+static
+int getCPUs(hwloc_cpuset_t cpuset, int *cpus, int size) {
+  int count = 0;
+  for (int i = hwloc_bitmap_first(cpuset);
+        i <= hwloc_bitmap_last(cpuset) && (count < size); i++) {
+    if (hwloc_bitmap_isset(cpuset, i)) {
+      cpus[count++] = i;
+    }
+  }
+  return count;
+}
+
+
+//
+// Fills the "cpus" array with up to "count" physical OS indices of the
+// accessible cores or PUs. If "physical" is true, then "cpus" contains
+// core indices, otherwise it contains PU indices. Returns the number
+// of indices in the "cpus" array.
+//
+int chpl_topo_getCPUs(chpl_bool physical, int *cpus, int count) {
+  return getCPUs(physical ? physAccSet : logAccSet, cpus, count);
```

Comment:
I think the ordering is subtle here. This requires that `chpl_topo_getNumCPUs{Physical,Logical}` functions are called before this function is called, which may be true in the current runtime but doesn't have to be. In the short term I think we should do a NULL check for `physAccSet`/`logAccSet`, and open up an issue to consider restructuring this.

Reply:
Although, maybe it's highly likely `chpl_topo_getNumCPUs*` is called first, since you need to know the number of cpus to call this function.

Reply:
I'll add the check and create an issue.

Reply:
New version creates a race. The way it's currently coded we should only call `getNumCPUs` with pthread_once. Let's just error if things are null, or skip the null check and do:

```c
CHK_ERR(pthread_once(&numCPUs_ctrl, getNumCPUs) == 0);
```

Reply:
Fixed.